### PR TITLE
Faster wallet refreshes

### PR DESCRIPTION
**Pull request by moneromooo-monero**
This has random stuff in it, because I need random stuff that's not merged, so all this has to be plonked in this tree for me to hack on it. I'll rebase once those things are merged.

Warning:
If you run this, your blockchain DB will upgrade to v2, which means you won't be able to run any daemon that does not have this patch (I'll get it merged once 0.12.1.0 is out of the way).

Speeds up between 2x and 5x. Speed tests welcome.

If your wallet is not a mining wallet (ie, you never got any coinbase output on it), setting refresh-type=no-coinbase now gives a pretty good speed boost.

vtnerd's wallet asm patch will give more speedup, as the wallet is using 90% of its time in the fe API.

### CODE REVIEW

**Comment by glv2** [comment]
Here are some test results of a wallet scanning the whole blockchain:

computer
 - OS: GNU/Linux x86-64
 - CPU: Intel Core i7-3630QM @ 2.40GHz
 - Storage: Samsung SSD 850 PRO

master branch (4b728d7dd48584987f53995a141baac4f886f017)
 - blocks: 1577488
 - time: 31.3 minutes

master branch + PR3716
 - blocks: 1577521
 - time: 8.6 minutes

master branch + PR3716 + refresh-type=no-coinbase
 - blocks: 1577526
 - time: 7.2 minutes

**Comment by moneromooo-monero** [comment]
Now ready for review.

**Comment by stoffu** [comment]
Computer:

- OS: Windows 7 Professional, 64bit
- CPU: Intel Core i7-5690X @ 3.00GHz
- Storage: Samsung SSD 850 PRO

Master (25e7a7d9):
- optimize-coinbase -> 26m 46s (1592555 blocks)
- no-coinbase -> 24m 24s (1592573 blocks)

This patch (c4fda6d0):
- optimize-coinbase -> 5m 31s (1592557 blocks)
- no-coinbase -> 4m 52s (1592578 blocks)

I can do two more tests on OSX machines with SSD and HDD.

**Comment by stoffu** [comment]
Note for testers: if you want to measure the time of scanning the entire blockchain from the genesis, not only the restore height has to be set to 0, but also the wallet creation time must not be set to current. That is, if you create a new wallet and `set refresh-from-block-height 0` and then `rescan_bc`, the wallet will NOT do the full scan of the blockchain (instead it skips old blocks based on timestamps). The correct way is to restore the wallet from seed and setting the restore height to 0.

**Comment by stoffu** [comment]
Computer:
- OS: Mac OSX 10.12.6
- CPU: 2.9 GHz Intel Core i7
- Storage: APPLE SSD SM2048L

Master:
- optimize-coinbase -> 23m 12s (1592682 blocks)
- no-coinbase -> 21m 31s (1592694 blocks)

PR-3716:
- optimize-coinbase -> 11m 9s (1592697 blocks)
- no-coinbase -> 9m 17s (1592702 blocks)

**Comment by stoffu** [comment]
Computer:
- OS: macOS Sierra 10.12.6
- CPU: 2.3 GHz Intel Core i7
- Storage: APPLE HDD HTS541010A9E662

Master:
- optimize-coinbase -> 1h 58m 24s (1592635 blocks)
- no-coinbase -> 1h 56m 42s (1593176 blocks)

PR-3716:
- optimize-coinbase -> 2h 0m 18s (1593251 blocks)
- no-coinbase -> 1h 57m 50s (1593310 blocks)

**Comment by moneromooo-monero** [comment]
That one seems odd since HDD means I/O is a bottleneck, and the new version reads much less from the database, since it stores prunable and unprunable data separately...

**Comment by stoffu** [comment]
OK, perhaps I might have made a mistake or something. I’ll test again.

**Comment by SamsungGalaxyPlayer** [comment]
I will test this on the new 970 EVO when I get a chance. I'll update this comment when done.

File: src/wallet/wallet2.h [hunk 1]
```diff
@@ -1282,6 +1295,8 @@ namespace tools
     std::string m_ring_database;
     bool m_ring_history_saved;
     std::unique_ptr<ringdb> m_ringdb;
+    boost::optional<crypto::chacha_key> m_ringdb_key;
+    unsigned int m_ringdb_key_refs;
```
Comment (stoffu) [inline src/wallet/wallet2.h]: Unused?

File: src/wallet/wallet2.cpp [hunk 2]
```diff
@@ -5791,6 +5810,7 @@ bool wallet2::get_ring(const crypto::chacha_key &key, const crypto::key_image &k
 
 bool wallet2::get_rings(const crypto::hash &txid, std::vector<std::pair<crypto::key_image, std::vector<uint64_t>>> &outs)
 {
+  key_ref kref(*this);
```
Comment (stoffu) [inline src/wallet/wallet2.cpp]: Is this needed?

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -2332,6 +2333,8 @@ bool wallet2::delete_address_book_row(std::size_t row_id) {
 //----------------------------------------------------------------------------------------------------
 void wallet2::refresh(uint64_t start_height, uint64_t & blocks_fetched, bool& received_money)
 {
+  key_ref kref(*this);
```
Comment (stoffu) [inline src/wallet/wallet2.cpp]: Is this needed?

Comment (moneromooo-monero) [reply src/wallet/wallet2.cpp]: No, but it speeds up saving rings from incoming txes if there are several.

**Pull request merged**